cpp/ql/src/semmle/code/cpp/exprs/Call.qll

```diff
@@ -134,6 +153,110 @@ abstract class Call extends Expr, NameQualifiableElement {
   }
 }
 
+private newtype TArgumentID =
+  TPositionalArgumentID(int n) {
+    exists(Function f |
+      n in [0 .. f.getNumberOfParameters() - 1]
+    )
+    or
+    exists(FunctionPointerIshType t |
+      n in [0 .. t.getNumberOfParameters() - 1]
+    )
+  }
+  or
+  TVarargsArgumentID(int n) {
+    exists(FunctionCall c |
+      n in [0 .. c.getNumberOfArguments() - c.getTarget().getNumberOfParameters() - 1]
+    )
+    or
+    exists(ExprCall c |
+      n in [0 .. c.getNumberOfArguments() - c.getExpr().getUnspecifiedType().(FunctionPointerIshType).getNumberOfParameters() - 1]
+    )
+  }
+  or
+  TQualifierArgumentID()
+
+
+/**
+ * An object representing a position within a call, either as a qualifier or as a
+ * positional argument.
+ */
+class ArgumentID extends TArgumentID {
+  /** Gets a textual representation of this argument ID. */
+  string toString() {
+    none()
+  }
+
+  predicate isQualifier() { none() }
+
+  predicate isPositional() {none() }
+
+  predicate isVararg() {none() }
+
+  int getArgIndex() { none() }
+}
+
+/**
+ * An argument identifier representing the qualifier.
+ */
+class QualifierArgumentID extends ArgumentID, TQualifierArgumentID {
+  override final predicate isQualifier() {
+    any()
+  }
+
+  override final string toString() {
+    result = "Arg(this)"
+  }
+}
+
+/**
+ * An argument identifier representing a specific positional argument index.
+ */
+class PositionalArgumentID extends ArgumentID, TPositionalArgumentID {
+  int argIndex;
+
+  PositionalArgumentID() {
+    this = TPositionalArgumentID(argIndex)
+  }
+
+  override final string toString() {
+    result = "Arg(" + argIndex + ")"
+  }
+
+  override final predicate isPositional() {
+    any()
+  }
+
+  /** Gets the index this argument identifier represents */
+  override final int getArgIndex() {
+    result = argIndex
+  }
+}
+
+/**
+ * An argument identifier representing a specific positional argument index.
+ */
+class VarargsArgumentID extends ArgumentID, TVarargsArgumentID {
+  int argIndex;
+
+  VarargsArgumentID() {
+    this = TVarargsArgumentID(argIndex)
+  }
+
+  override final string toString() {
+    result = "Vararg(" + argIndex + ")"
+  }
+
+  override final predicate isVararg() {
+    any()
+  }
+
+  /** Gets the index this argument identifier represents */
+  override final int getArgIndex() {
+    result = argIndex
```

Comment:
I think it's too error-prone to make `getArgIndex` a "virtual" predicate. It doesn't make sense to call it without knowing the subtype of `ArgumentID` on which you're calling it. That's evident in `getArgumentByID`, its sole caller.